lib/src/config_concatenation.cc

```diff
@@ -0,0 +1,248 @@
+                }
+            }
+
+            vector<shared_value> consolidated;
+            consolidated.reserve(flattened.size());
+
+            for (auto& v : flattened) {
+                if (consolidated.empty()) {
+                    consolidated.push_back(v);
+                } else {
+                    join(consolidated, v);
+                }
+            }
+
+            return consolidated;
+        }
+    }
+
+    shared_value config_concatenation::concatenate(std::vector<shared_value> pieces) {
+        vector<shared_value> consolidated { consolidate(pieces) };
+        if (consolidated.empty()) {
+            return nullptr;
+        } else if (consolidated.size() == 1) {
+            return consolidated.front();
+        } else {
+            shared_origin merged_origin = simple_config_origin::merge_origins(consolidated);
+            return make_shared<config_concatenation>(merged_origin, consolidated);
+            return {};
+        }
+    }
+
+    // when you graft a substitution into another object,
+    // you have to prefix it with the location in that object
+    // where you grafted it; but save prefixLength so
+    // system property and env variable lookups don't get
+    // broken.
+    shared_value config_concatenation::relativized(std::string prefix) const {
+        vector<shared_value> new_pieces;
+        for (auto& p : _pieces) {
+            new_pieces.push_back(p->relativized(prefix));
+        }
+
+        return make_shared<config_concatenation>(origin(), move(new_pieces));
+    }
+
+    bool config_concatenation::operator==(config_concatenation const& other) const {
+        // note that "origin" is deliberately NOT part of equality
+        return (_pieces == other._pieces);
+    }
+
+    shared_value config_concatenation::new_copy(shared_origin origin) const {
+        return make_shared<config_concatenation>(move(origin), _pieces);
+    }
+
+    bool config_concatenation::ignores_fallbacks() const {
+        // we can never ignore fallbacks because if a child ConfigReference
+        // is self-referential we have to look lower in the merge stack
+        // for its value.
+        return false;
+    }
+
+    void config_concatenation::render(std::string& result, int indent, bool at_root, config_render_options options) const {
+        for (auto& p : _pieces) {
+            p->render(result, indent, at_root, move(options));
+        }
+    }
+
+    config_exception config_concatenation::not_resolved() const {
+        return config_exception("need to config#resolve(), see the API docs for config#resolve; substitution not resolved");
+    }
+
+    bool config_concatenation::is_ignored_whitespace(shared_value value) {
+        auto value_string = dynamic_pointer_cast<const config_string>(value);
+        return (value_string && !(value_string->was_quoted()));
+    }
+
+    void config_concatenation::join(std::vector<shared_value> & builder, shared_value right) {
```

Comment:
Might be nice to bring over the comment: `Add left and right, or their merger, to builder.`